File: day018/128. 最长连续序列.py
class Solution:
    def longestConsecutive(self, nums: List[int]) -> int:
        # 时间复杂度O(NlogN)，空间复杂度O(1)
        # 快排
        if not nums:
            return 0
        nums.sort()
        temp = 1
        long = 1
        for i in range(1, len(nums)):
            if nums[i] == nums[i-1]:
                continue
            if nums[i] == nums[i-1] + 1:
                temp += 1
                long = max(long, temp)
            else:
                temp = 1
        return long
                
    def longestConsecutive(self, nums: List[int]) -> int:
        # 时间复杂度O(N)，空间复杂度O(N)
        # 哈希表
        if not nums:
            return 0
        se = set(nums)
        temp = 1
        long = 1
        for num in nums:
            if num - 1 not in se:
                while num + 1 in se:
                    temp += 1
                    long = max(temp, long)
                    num += 1
                temp = 1
        return long
            

File: day018/128. 最长连续序列.cpp
class Solution {
public:
    int longestConsecutive(vector<int>& nums) {
        unordered_set<int> myset(nums.begin(), nums.end());
        int curLen = 0, maxLen = 0;
        for (int num : nums)
        {
            if (myset.count(num - 1) == 0)
            {
                curLen = 1;
                while (myset.count(num + 1))
                {
                    curLen++;
                    num++;
                }
                maxLen = maxLen > curLen ? maxLen : curLen;
            }
        }
        return maxLen;
    }
};
